### ISSUE TITLE
kubectl annotate breaks slashes in keys with cli 1.7 and master 1.8

### ISSUE THREAD

**Issue by thomas-riccardi** [kubernetes/kubernetes#59200]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
`kubectl annotate` is broken when annotation key contains a slash with cli 1.7 and master 1.8. 

**What you expected to happen**:
`kubectl annotate` should work even with slashes in annotation keys (it's allowed) with one minor version skew between cli and master (see https://github.com/kubernetes/website/issues/7103).

**How to reproduce it (as minimally and precisely as possible)**:
Use cli 1.7.12 and master 1.8.5 (probably same issue with other master 1.8 versions)
```
$ kubectl create configmap my-config --from-literal=key1=config1
configmap "my-config" created

 $ kubectl annotate configmap my-config example.com/foo=bar
The ConfigMap "my-config" is invalid: metadata.annotations: Invalid value: "example.com~1foo": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
```

Logs with `-v 10`:
```
I0201 11:13:19.813961    2540 request.go:991] Request Body: {"metadata":{"annotations":{"example.com~1foo":"bar"}}}
```

**Anything else we need to know?**:

Some other tests:
- OK `foo=bar`
- OK `test/foo=bar`
- KO `test.com/foo=bar`
- KO `example/foo=bar`
- KO `teste/foo=bar`

This is strange: 
- `test/` prefix seems special: it's the only one accepted 
- the regex is unexpected: aren't the slashes allowed? They are used everywhere in kubernetes.

It all works with the same master and the cli `1.8.5` and `1.9.2`.

This has already been bug reported but dismissed as `probably related to using incompatible versions` (see https://github.com/kubernetes/kubectl/issues/60 and #52283). But cli 1.7 should work with master 1.8, shouldn't it?

**Environment**:
- Kubernetes version (use `kubectl version`):
```$  kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.12", GitCommit:"3bda299a6414b4866f179921610d6738206a18fe", GitTreeState:"clean", BuildDate:"2017-12-29T09:00:00Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.5-gke.0", GitCommit:"2c2a807131fa8708abc92f3513fe167126c8cce5", GitTreeState:"clean", BuildDate:"2017-12-19T20:05:45Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
  - cluster: GKE
  - client: GCE
- OS (e.g. from /etc/os-release):
ubuntu 16.04

**Comment by yue9944882** [comment]
@thomas-riccardi 
It should have fixed in `evanphx/json-patch` after [https://github.com/evanphx/json-patch/commit/c3740aa533c8ac9463b1428e3f964f66733135af](https://github.com/evanphx/json-patch/commit/c3740aa533c8ac9463b1428e3f964f66733135af).
An update of that vendor dependency is needed for aligning revision to release-1.8 or later.

**Comment by yue9944882** [comment]
Fixed by #59324
/close

**Comment by ajit-hybris** [comment]
I have the same issue with k8s 1.12.1
